src/pages/authentication/PickNamePage.tsx

```diff
@@ -10,53 +10,63 @@ import {
 import MinimalWrapper from '../../layout/MinimalLayout/MinimalWrapper';
 import CenteredLayout from './CenteredLayout';
 import SnackbarAlert from './SnackbarAlert';
-
-const API = '/data-api/rest/volunteer';
-const HEADERS = {
-  Accept: 'application/json',
-  'Content-Type': 'application/json;charset=utf-8',
-};
-
-interface Volunteer {
-  id: number;
-  name: string;
-}
+import { getRole, UserContext } from '../../components/contexts/UserContext';
+import { ENDPOINTS, HEADERS } from '../../types/constants'
+import { Volunteer } from '../../types/interfaces';
+import { IdTokenClaims } from '@azure/msal-common';
+import { useMsal } from '@azure/msal-react';
 
 const PickYourNamePage: React.FC = () => {
   const [selectedVolunteer, setSelectedVolunteer] = useState<Volunteer | null>(
     null,
   );
   const [volunteers, setVolunteers] = useState<Volunteer[]>([]);
   const [openSnackbar, setOpenSnackbar] = useState<boolean>(false);
-
+  const [user, setUser] = useState<IdTokenClaims | null>(null);
   const navigate = useNavigate();
+  const { instance } = useMsal();
 
   useEffect(() => {
+    const account = instance.getActiveAccount();
+    if (account) {
+      instance
+        .acquireTokenSilent({
+          account: account,
+          scopes: ['openid', 'profile', 'email', 'User.Read'],
+        })
+        .then((response) => {
+          const userClaims = response.idTokenClaims;
+          setUser(userClaims || null);
```

Comment:
From my understanding, the `setUser` function is used to store user information in the global User context, such as the user's role. This is important because line 51, `HEADERS['X-MS-API-ROLE'] = getRole(user);`, depends on retrieving the role correctly. 

Currently, both the admin and volunteer are calling `setUser` in the `mainlayout/index` file on line 47 with `setUser(userClaims || null);`, which means that the volunteer is calling `setUser` twice which is  inefficient. I think it can be optimized in the future. Perhaps we could create a task in Jira to address this issue.

Reply:
@pingchenchan logged[ #193](https://das-ph-inventory-tracker.atlassian.net/browse/PIT-193)